### DISCUSSION TITLE
Issue with next-multilingual package and Next.js

### DISCUSSION THREAD

**Discussion by nbouvrette** [vercel/vercel#6710] [General] [answered]
We recently released a brand new i18n package (https://github.com/Avansai/next-multilingual) and it requires reading from the build manifest to generate the SSR localized URLs. This works fine locally (tested on OSX, Linux, Windows). On Vercel it also works 99%, for example, if you hit https://next-multilingual-example.vercel.app/en-us/about-us it will render and work. The problem is that when you hit `/` it returns a 404, but if you inspect the console it actually returns a 500 with the following error in the Vercel logs:

```
Error: ENOENT: no such file or directory, open '/var/task/.next/static/CREmo4Fl-BYNvY7GSUx2t/_buildManifest.js'
```

Basically, we were planning to use our own package on our main project and now it's blocking our backlog. The team discussed doing our own hosting but I feel we might be losing some of the optimizations Vercel did.

How much time should we give on our existing support ticket for that type of issue, or should we just go directly to the hosting options?

**Comment by leerob** [comment]
To confirm, you're on a Pro plan hosting with Vercel? What is your team name? Paid customers are first in the queue (which makes sense) so I just wanted to check 😄 

> How much time should we give on our existing support ticket for that type of issue

It's impossible to say how long something will take to resolve without investigating the issue first. Once we investigate, we can provide an update!

Hope this helps.

**Reply by nbouvrette** [reply]
Yes, we are on a pro plan but have been getting responses every few days - our team is called Avansai (https://vercel.com/avansai)

I was thinking of giving a maximum of 2 weeks before moving to another host since it's blocking our entire product release but we would really prefer to use Vercel.

If you have any tricks to help us, we are happy to even jump on a call to do live troubleshooting.

Thanks!

**Reply by leerob** [reply]
I took a look at your case, where it appears y'all sent over the source code for us to take a look. Thank you! Sorry for the response there. It appears it's a library-specific issue with the package and not a Next.js specific bug. Further, you mentioned it's looking at undocumented APIs like `_buildManifest`, which is likely not the best solution.

Have you verified this package is working when deployed with `next start`? You might also consider looking at other open source i18n solutions to see how they're constructed. Finally, you might be able to get some assistance from the community in the Next.js GitHub Discussions as well.

**Reply by nbouvrette** [reply]
We wrote this new i18n library to hopefully fill a few gaps that are not covered by Next.js - this originated from this discussion https://github.com/vercel/next.js/discussions/20240.

We know using `_buildManifest` is not standard but it works with `next start`. We also tested this on all OSes and it works well, so it does seem like some sort of issue with how Vercel hosts the app.

We also got a demo up fully working on Google cloud here which is also why we are considering other hosting options: https://next-multilingual-example.avansai.com/

This is also why I wanted your input since I know this is probably not a standard support request and we cannot wait indefinitely :) Our ultimate goal would be to bring some of the new ideas of our i18n library into Next.js natively but our first step was to illustrate how it could work and get feedback from the community. Of course, it would be a lot better if it would work with Vercel!

**Reply by leerob** [reply]
Thanks for contributing back to the ecosystem, I checked out the package and it's very well documented 😄 Are you taking advantage of reading/writing to the FS during the build phase or at runtime? It's possible that's what causing an issue in serverless environments.

We'd love to support it once we can pinpoint what the library is doing, but understood if you need to host serverful while we figure that out!

**Reply by nbouvrette** [reply]
Its been a lot of effort which is why it would be a shame to not get this supported by Vercel :) Our private repo is already all set up on Vercel but we cannot start localizing it with our new library unless we change host right now...

Regarding using `fs`, we only use it at build time. If you check the error in the Vercel console:

```
 at getRewrites (/var/task/.next/server/chunks/18.js:2092:62)
    at MulLink (/var/task/.next/server/chunks/18.js:2172:107)
```

It looks like our SSR localized links are incompatible with Vercel. `getRewrites` is only used in this file https://github.com/Avansai/next-multilingual/blob/main/src/link/ssr.tsx and this file is only used in Webpack in the `next.config.js` like this:

```
    webpack(config, { isServer }) {
        if (isServer) {
            config.resolve.alias['next-multilingual/link$'] = require.resolve('next-multilingual/link/ssr');
        }
        return config;
    }
```

If you think we can work this together I am willing to stick with Vercel as long as we can! Thanks for your help so far

**Reply by nbouvrette** [reply]
@leerob one other thing, is that our `/` is using `getServerSideProps` to be able to dynamically render the correct locale without redirecting. For some odd reason, this is the only page that is not working on Vercel - so I wonder if there isn't some other problem specific to `/`

**Comment by styfle** [comment, marked as answer]
The issue is the `next-multilingual` plugin is trying to read client-side JS from the server-side JS (aka the serverless function) here:

https://github.com/Avansai/next-multilingual/blob/96676021ea1c4d43739ce5d03395ebf46d1a497f/src/helpers/get-rewrites.ts#L26

Instead, that plugin should do `fs.readFile('.next/routes-manifest.json')` which is available after `next build`.

**Reply by nbouvrette** [reply]
@styfle so basically none of the files in the `static` folder are available when doing `getServerSideProps`?

If `routes-manifest.json` is available I think it should be possible to push a fix over the weekend and test it out. Thanks for explaining this mysterious behavior, we really appreciate it!

Will keep you posted as soon as we test a fix.

**Reply by nbouvrette** [reply]
@styfle I confirm that by adding `routes-manifest.json` the issue is resolved. Thanks for taking the time to explain why it didn't work on Vercel while it worked everywhere else! This file seems only available on builds which is also why we didn't use it previously. If you are curious:

Code fix: https://github.com/Avansai/next-multilingual/commit/cf3d68d1b0dec78bcea58b4c615c9cd80c015557
Working app: https://next-multilingual-example.vercel.app/

Special thanks to @leerob for helping to escalate this and also the person who was also communicating with us by email. 

Great customer service and kudos to Vercel!
